keps/sig-network/1880-services-ip-ranges/README.md

```diff
@@ -0,0 +1,1030 @@
+c.recorder.Eventf(&svc, nil, v1.EventTypeWarning, "ClusterIPOutOfRange", "ClusterIPAllocation", "Cluster IP [%v]:%s is not within the service CIDR %s; please recreate service",
+```
+
+- Valid Service and ServiceIPRange but not IPAddress
+
+It can happen that an user forcefully delete an IPAddress, in this case, the controller will regenerate the IPAddress.
+
+During this time, there is a chance that an apiserver tries to allocate this IPAddress, with a possible situation where
+2 Services has the same IPAddress. In order to avoid it, the Allocator will not delete an IP from its local cache
+until it verifies that the Service associated has been deleted too.
+
+- Valid IPAddress and ServiceIPRange but no Service
+
+The IPAddress will be deleted and an event generated if the controller determines that the IPAddress
+is orphan [(see Allocator section)](#Allocator)
+
+#### Resizing Service IP Ranges
+
+One of the most common problems users may have is how to scale up or scale down their service IP ranges.
+
+Let's see an example on how these operations will be implemented.
+
+Assume we have  configured a service range 10.0.0.0/24 and its fully used, we can:
+
+1. Add another /24 range 10.0.1.0/24 and keep the previous one
+2. Add an overlapping larger range 10.0.0.0/23
+
+The user can now remove the first /24 range, since the new range covers all the existing IP Addresses
+
+The same applies for scaling down, meanwhile the new range contains all the existing IPAddresses, the old
+range will be safely deleted.
+
+However, there can be a race condition during the operation of scaling down, since the propagation of the
+deletion can take some time, one allocator can successfully allocate an IP address out of new range (but still
+inside of the old range).
+
+There is one controller that will periodically check that 1on1 relation between IPAddresses and Services is
+correct, and will start sending events to warn the user that it has to fix/recreate the corresponding Service.
+
+
+
+#### API
+
+
+```go
+// ServiceIPRange defines a range of IPs using CIDR format (192.168.0.0/24 or 2001:db2::0/64).
+type ServiceIPRange struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ServiceIPRangeSpec   `json:"spec,omitempty"`
+}
+
+
+// ServiceIPRangeSpec describe how the ServiceIPRange's specification looks like.
+type ServiceIPRangeSpec struct {
+	// CIDR is an IP block in CIDR notation "10.0.0.0/8" or "fd12:3456:789a:1::/64"
+  // in using its canonical representation (rfc5952)
+	CIDR string `json:"cidr"`
+}
+
+// IPAddress represents an IP used by Kubernetes associated to an ServiceIPRange.
+// The name of the object is the IP address in canonical format.
+type IPAddress struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   IPAddressSpec   `json:"spec,omitempty"`
+}
+
+// IPAddressSpec describe the attributes in an IP Address,
+type IPAddressSpec struct {
```

Comment:
I guess we need our own ref

Reply:
what about https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#multiple-resource-reference

```go
serviceRef:
    apiVersion: v1
    kind: Service
    name: kubernetes
    namespace: default
```

I really love if we can set this reference during the allocation, in BeginCreate, it will simplify a lot of problems like detecting leaks, ...

Reply:
I think here we need to use "resource" rather thsan "kind" because we don't know it, and the kind could be presented on multiple resources.  Sigh.

Reply:
I took the same one that Gateway uses, group and kind should be enough, no?
```
type ParentReference struct {
  // Group is the group of the referent.
  Group string
  // Kind is kind of the referent.
  Kind string
  // Namespace is the namespace of the referent
  Namespace string
  // Name is the name of the referent
  Name string
}
```

Reply:
It can't be group/kind because we can't assert that the controller knows about the kinds, which is required to use "kind".  Gateway impls *do* know each kind, so it's OK.   Also, Gateway is a user-facing API, this is not, so it's not as ugly.

Reply:
Ic, it is well explained here https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#kind-vs-resource

resource it is

Reply:
also added UID